### DISCUSSION TITLE
What is limited in SQLite support?

### DISCUSSION THREAD

**Discussion by dmi-dmi-dmi** [piccolo-orm/piccolo#624] [Q&A] [answered]
[Database Support](https://piccolo-orm.readthedocs.io/en/latest/piccolo/getting_started/database_support.html) section in the docs states that SQLite support is limite in Piccolo.

Could someone please clarify what exactly is limited in comparison with Postgres?

I want to implement a basic desktop software in Python so using SQLite is somewhat natural in my case so I want to know if I still can use Piccolo.

Thanks

**Comment by dantownsend** [comment, marked as answer]
Hi @dmi-dmi-dmi 

Almost all features are available for SQLite. The main limitation is Piccolo's migrations. SQLite has very limited `ALTER TABLE` support, which makes it hard for us to support it with migrations in Piccolo.

So when you're creating your database tables, you'll have to write a script to do it manually. Something like this:

```python
from piccolo.table import create_db_tables_sync

create_db_tables_sync(TableA, TableB, TableC)
```

The docs are available [here](https://piccolo-orm.readthedocs.io/en/latest/piccolo/query_types/create_table.html#create-db-tables-create-db-tables-sync).

**Reply by Conobi** [reply]
Hey! I was tinkering with migrations with SQLite, and building a custom migration is quite painful compared to the awesome auto migration tool you built.
Just for fun I tried to disable the check of SQLite engine on `--auto`, and building, running forward/backward auto migrations on my SQLite setup (in `3.43.1` version) seemed to work well.
Precisely, what are the caveat that could go \*really wrong\* with doing this on a SQLite engine?
Since SQLite is gaining popularity (and Piccolo too), wouldn't it be possible to support auto migrations on SQLite?

Thanks!

**Reply by dantownsend** [reply]
It's a good point - I know that SQLite has been getting better over time with its support for `ALTER TABLE` statements.

I just checked, and Python 3.12 on my laptop now has SQLite 3.42.0.

It looks like the features now supported in SQLIte are:

* `DROP COLUMN` was added in [3.35.0](https://www.sqlite.org/releaselog/3_35_0.html)
* `RENAME COLUMN` was added in [3.25.0](https://www.sqlite.org/releaselog/3_25_0.html)
* `ADD COLUMN` was added in [3.2.0](https://sqlite.org/changes.html#version_3_2_0)
* `RENAME TABLE` was added in [3.1.0](https://www.sqlite.org/releaselog/3_1_0.html)

So in recent-ish versions, it's just altering columns which is the problem (e.g. making a column nullable).

The work around in SQLite has always been to do the following:

* Create a new table
* Copy the old data into the new table
* Drop the old table
* Rename the new table

It always felt like a hack, and I was hoping that at some point SQLite would gain full support for `ALTER TABLE`, but I don't think it will happen soon. So I think we'll just have to add support for the work around, and use that. 

I agree that we should just remove this:

https://github.com/piccolo-orm/piccolo/blob/5f972d0a7af45aeb3b18fc02f61b8c913f03b118/piccolo/apps/migrations/commands/new.py#L234-L235

At the very worst it should output a warning, or ask the user if they want to continue, because the auto migrations will work with SQLIte in most situations.

**Reply by Conobi** [reply]
Yes! A warning to tell to the user that it could break and to always test the migration is definitely a good idea. An explanation in the docs on what could break would be nice too, even if it is just with a link to the ALTER page of SQLite docs
